### ISSUE TITLE
Cross domain problems with raster source

### ISSUE THREAD

**Issue by jasonwoo** [mapbox/mapbox-gl-js#2171]
Can anyone confirm if something like this should work using external raster tiles? I'm getting: "No 'Access-Control-Allow-Origin' header is present on the requested resource."

```
style = {
"version": 8,
"sources": {
            "raster-tiles": {
                "type": "raster",
                "tiles": ['https://some-external-server.com/tiles/{z}/{x}/{y}.jpg'],
                "tileSize": 256
            }
},
"layers": [{
            "id": "test-tiles",
            "type": "raster",
            "source": "raster-tiles",
            "minzoom": 0,
            "maxzoom": 10
        }]
};
```

```
map = new mapboxgl.Map({
    	 container: 'map', 
    	 style: style,
    	 center: [-75.13,38.79], 
});
```
Thanks!

**Comment by tmcw** [comment]
> No 'Access-Control-Allow-Origin' header is present on the requested resource.

This is a descriptive error: does your server [support CORS](http://enable-cors.org/)?

**Comment by jasonwoo** [comment]
Thanks for the reply. It's not my server but I would say it's probably unlikely that it would support CORS.  
I'm currently using these tiles with "L.mapbox.tileLayer" in the Mapbox JS API and they load ok. I was just seeing what it would take to move things over to the GL JS API and ran into that. I thought (probably incorrectly) that I could add them as a raster source and things would behave the same way.

**Comment by ansis** [comment]
Unfortunately we can't support raster sources without CORS.

To use the raster images in WebGL we need to access the data in the images programmatically. We can only do this if the server supports CORS. Leaflet doesn't need this because it never accesses the data in the images. It just moves the `<img>`s around.

**Comment by imran50** [comment]
I am facing the problem of CORS.. want to use WMS Geoserver request with Mapbox GL and facing the problem of CORS. 
just want to clear the concept of the request, Is it the problem with the Geoserver or client-side Mapbox GL?
what should i improve to solve this problem Mapbox GL or Geoserver ?
Thanks

**Comment by jasonwoo** [comment]
WMS layers should work in the latest Mapbox GL:

https://www.mapbox.com/mapbox-gl-js/example/wms/

My guess is you need to enable CORS in your Geoserver instance.

**Comment by HarelM** [comment]
Is there any workaround for this?
When using openlayers, which uses a similar technique (a canvas) there's no CORS problem for external tile servers.
I'm in the middle of migrating from openlayers to this library when I ran into this issue. Needless to say this is a showstopper as my site allows adding external tile servers.

**Comment by risharde** [comment]
I'm having this issue even using this official example from mapbox (which is rather ironic)
https://docs.mapbox.com/mapbox-gl-js/example/add-terrain/

I get the cors warning as well:
"...has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled."

Any ideas how I can get around this issue?

**Comment by kieranongh** [comment]
@risharde You could set up a proxy
Turns out there's a few services out there for development (ie. [crossorigin.me](https://corsproxy.github.io/) and many frontend frameworks have proxy features [React](https://create-react-app.dev/docs/proxying-api-requests-in-development/)

Pretty standard to make proxies for production environments as well.

I'm not sure if there's any caveats of doing it this way, but it seems pretty common using proxies in production

**Comment by ansis** [comment]
@risharde it looks like that is an issue with the api responses. I've notified the team responsible for it and they are looking into it. Thanks for reporting it!

We sometimes miss new responses on issues that have been already closed for a long time. Opening a new issue or contacting mapbox support can be a good thing to do for these kinds of errors.

**Comment by risharde** [comment]
Thanks guys, as you can see, I've been waiting a month and a half on this, so I'm hoping this would be soon since this should be an easy fix?

**Comment by csimpi** [comment]
When are you guys going to fix this? Being unable to load an image from S3 is an issue that you guys should be able to handle in some way.

**Comment by AliAkinK** [comment]
Eight years later, Mapbox themselves haven't even fixed their own examples or self hosted maps lol

Open up a preview or an embed URL from one of your maps and check the console log -- ridiculous.
